When encountering `...` when `..` is expected, parser should suggest the latter

Consider the following code:

```rust
fn main() {
    struct S { one: u32, _two: u32 }
    match (S { one: 1, _two: 2 }) {
        S { one: 1, ... } => println!("oy"),
        _ => println!("no"),
    }
}
```

If you try to compile this, you get the following errors:

```
error: expected identifier, found `...`
 --> src/main.rs:4:21
  |
4 |         S { one: 1, ... } => println!("oy"),
  |                     ^^^

error[E0027]: pattern does not mention field `one`
 --> src/main.rs:4:9
  |
4 |         S { one: 1, ... } => println!("oy"),
  |         ^^^^^^^^^^^^^^^^^ missing field `one`

error[E0027]: pattern does not mention field `_two`
 --> src/main.rs:4:9
  |
4 |         S { one: 1, ... } => println!("oy"),
  |         ^^^^^^^^^^^^^^^^^ missing field `_two`

error: aborting due to 3 previous errors
```

See the bug?

Its because I used `...` when I should have used `..`

It would be nice if the compiler, in its ever friendly manner, suggested the latter to me.

--- a/src/libsyntax/parse/parser.rs
+++ b/src/libsyntax/parse/parser.rs
@@ -1212,6 +1212,9 @@ impl<'a> Parser<'a> {
     pub fn span_err(&self, sp: Span, m: &str) {
         self.sess.span_diagnostic.span_err(sp, m)
     }
+    pub fn struct_span_err(&self, sp: Span, m: &str) -> DiagnosticBuilder<'a> {
+        self.sess.span_diagnostic.struct_span_err(sp, m)
+    }
     pub fn span_err_help(&self, sp: Span, m: &str, h: &str) {
         let mut err = self.sess.span_diagnostic.mut_span_err(sp, m);
         err.help(h);
@@ -3445,7 +3448,16 @@ impl<'a> Parser<'a> {
             let lo = self.span;
             let hi;
 
-            if self.check(&token::DotDot) {
+            if self.check(&token::DotDot) || self.token == token::DotDotDot {
+                if self.token == token::DotDotDot { // Issue #46718
+                    let mut err = self.struct_span_err(self.span,
+                                                       "expected field pattern, found `...`");
+                    err.span_suggestion(self.span,
+                                        "to omit remaining fields, use one fewer `.`",
+                                        "..".to_owned());
+                    err.emit();
+                }
+
                 self.bump();
                 if self.token != token::CloseDelim(token::Brace) {
                     let token_str = self.this_token_to_string();

--- a/src/test/ui/did_you_mean/issue-46718-struct-pattern-dotdotdot.rs
+++ b/src/test/ui/did_you_mean/issue-46718-struct-pattern-dotdotdot.rs
@@ -1,0 +1,27 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![allow(unused)]
+
+struct PersonalityInventory {
+    expressivity: f32,
+    instrumentality: f32
+}
+
+impl PersonalityInventory {
+    fn expressivity(&self) -> f32 {
+        match *self {
+            PersonalityInventory { expressivity: exp, ... } => exp
+            //~^ ERROR expected field pattern, found `...`
+        }
+    }
+}
+
+fn main() {}

--- a/src/test/ui/did_you_mean/issue-46718-struct-pattern-dotdotdot.stderr
+++ b/src/test/ui/did_you_mean/issue-46718-struct-pattern-dotdotdot.stderr
@@ -1,0 +1,8 @@
+error: expected field pattern, found `...`
+  --> $DIR/issue-46718-struct-pattern-dotdotdot.rs:21:55
+   |
+21 |             PersonalityInventory { expressivity: exp, ... } => exp
+   |                                                       ^^^ help: to omit remaining fields, use one fewer `.`: `..`
+
+error: aborting due to previous error
+

